crates/sequencer/src/state.rs

```diff
@@ -14,28 +17,58 @@ use starknet_api::{
     hash::StarkFelt,
 };
 
+use serde::{Deserialize, Serialize};
+
 use crate::commit::Committer;
+use crate::serde::{SerializableState, SerializationError};
 
 /// Generic state structure for the sequencer.
 /// The use of `FxHashMap` allows for a better performance.
 /// This hash map is used by rustc. It uses a non cryptographic hash function
 /// which is faster than the default hash function. Think about changing
 /// if the test sequencer is used for tests outside of ef-tests.
 /// See [rustc-hash](https://crates.io/crates/rustc-hash) for more information.
-#[derive(Default)]
+#[derive(Default, Debug, Serialize, Deserialize, PartialEq)]
 pub struct State {
-    classes: FxHashMap<ClassHash, ContractClass>,
-    compiled_class_hashes: FxHashMap<ClassHash, CompiledClassHash>,
-    contracts: FxHashMap<ContractAddress, ClassHash>,
-    storage: FxHashMap<ContractStorageKey, StarkFelt>,
-    nonces: FxHashMap<ContractAddress, Nonce>,
+    pub classes: FxHashMap<ClassHash, ContractClass>,
+    pub compiled_class_hashes: FxHashMap<ClassHash, CompiledClassHash>,
+    pub contracts: FxHashMap<ContractAddress, ClassHash>,
+    pub storage: FxHashMap<ContractStorageKey, StarkFelt>,
+    pub nonces: FxHashMap<ContractAddress, Nonce>,
 }
 
 impl State {
     /// Helper function allowing to set the nonce of a contract.
     pub fn set_nonce(&mut self, contract_address: ContractAddress, nonce: Nonce) {
         self.nonces.insert(contract_address, nonce);
     }
+
+    /// This will serialize the current state, and will save it to a path
+    pub fn dump_state_to_file(&self, path: &PathBuf) -> Result<(), SerializationError> {
```

Comment:
Otherwise, if the reason why you implemented From<&State> because of this, can't you just have `self` instead of `&self`?

Reply:
I think we should avoid passsing self as much as possible right?

Because, I think it moves a value { or you clone it }.

Anyway, I have changed the `From` trait implementation.

Reply:
I think self is ok if you need to and it removes a clone but in this case I think we can't avoid clone anyway

Reply:
Gotcha, lemme refactor then!